packages/modeling/test/helpers/compareVectors.js

```diff
@@ -1,10 +1,12 @@
+const { EPS } = require('../../src/maths/constants')
 /**
  * Compare two vectors for equality
- * @param (vec} vec1 - vector (array) of values
- * @param (vec} vec2 - vector (array) of values
+ * @param {vec} vec1 - vector (array) of values
+ * @param {vec} vec2 - vector (array) of values
+ * @param {number} eps - Epsilon - the largest difference between two numbers to consider trivial
  * @returns {boolean} result of comparison
  */
-const compareVectors = (vec1, vec2, eps = Number.EPSILON) => {
+const compareVectors = (vec1, vec2, eps = EPS) => {
```

Comment:
Indeed.  Number.EPSILON is intended to be the smallest difference for 2 numbers to be functionaly equal in the context of JS math.  The math/contants.js (I assume) is the smallest difference for two numbers to be equal in the context of jscad modelling.  MeasureBounds was returning differences 10x greater than number.EPSILON, but still many orders of magnitude less than EPS.  Since EPS is supposed to be the 'resolution' of jscad, this struck me as a sane change.

Reply:
Sadly, EPS has various meanings throughout the code, but recently there’s a more discrete definition...

EPS is the maximum distance between planes when comparing for equality.

In other words, P1 and P2 are the same if the distance between planes is less then EPS, and P1 and P2 are parallel. This is used inside the boolean operations, specifically the BSP trees.

Reply:
Can the same not be used for points?  Points are considered equal if they are with EPS distance of each other?

I guess we essentially have 2 discussions going on here.

- What do we use to define floating point equality in modelling?
- In the context of the test, what should we do if geometry results that should be equal are 10e-15 apart, and the epsilon we are using from Number is 10e-16?

Reply:
I think that you are understanding the issues well, which makes the discussions go quickly. Super.

Yeah. We definitely need to keep the testing independent of the code. Everything was working fine with the Number.EPSILON until recently. Then, I introduced the optional eps value on compareVectors() as some of the matrix math was returning results outside that EPSILON.

Hmm... I think comparePoints() could use a different epsilon value, but not sure about compareVectors.

Reply:
comparePoints just iteratively calls compareVectors.  We have, I suspect, a situation of compounded floating point arithmetic drift, propogated through several calculations.  Perhaps 10e-5 is overkill.  10e-16 is too small. 10e-14 would be sufficient, but I would suggest maybe 10e-10 or 10e-12 is a fairly sane compromise.  Anyone modelling on that small a scale could have no expectation of accuracy, I would think.

Reply:
That sounds fine.

I think there will be some changes in test suites in the future, so a little compromise now should be fine.